wiki/Operator-101.md

```diff
@@ -0,0 +1,124 @@
+# Contrail Operator 101
+
+## What is an operator?
+
+Operator is an automated software extension which allow to easily manage applications and its' components.<br/>
+It's used to package, deploy and manage applications based on Kubernetes custom resources.
+
+Operator acts as a controller which extend Kubernetes API to manage lifecycle of dependent application resources. Controller implements logic which periodically compare desired state of a cluster to it's actual state and apply corrections to meet declared state.
+
+Compared to other methods of managing application deployment on Kubernetes clusters, operator allows to cover completly lifetime of application.<br/>
+Below graph compares operator to Helm and Ansible.<br/>
+Helm provides only installation process with some mechanisms of upgrade.
+Ansible additionaly allow to cover some aspects of managing application resources lifecycle.<br/>
+However, only operator allows to have full insights into created cluster resources and perform custom operations on them during lifecycle.
+
+![operator-compared](images/operator-compared.png)
+
+Custom operator may be written in multiple languages as in fact it's just a piece of code that periodically act with requests to Kubernetes API.<br/>
+Currently, the most popular language to write operators is the language that was used to write Kubernetes project itself - Go.<br/>
+Helpful tool for creating operators is [Operator Framework](https://github.com/operator-framework) which distributes [Operator-SDK](https://github.com/operator-framework/operator-sdk) commonly used to create operators - for example this operator.
+
+Every custom resource built with operator contains 2 elements.<br/>
+First element is API which defines how resource is defined and what's it's structure.
+This definition is used afterwards by Operator Framework to generate CRD manifests applied on a cluster and defines how user should write manifests to deploy custom resource successfully.<br/>
+Second element is controller which runs on an operator pod and handles logic of custom resource.
+Every controller has Reconcile method which is run periodically and contain code which defines what to do every loop.
```

Comment:
Instead of "periodically" I would write that each reconciler uses k8s watch api to react to some dependant resources' changes (this is not active waiting). For example in case of Postgres these are: Postgres CR, k8s Pod, k8s Service. Each time one of these resources changes, Reconcile is triggered.

Reply:
Done